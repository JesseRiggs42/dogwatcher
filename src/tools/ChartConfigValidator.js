import {
    assertIsBoolean,
    assertIsDefinedNotNull,
    assertIsNumber,
    assertIsNumericArray,
    assertIsString,
    assertIsStringArray,
    assertIsSupported,
    assertIsValid,
    assertNonEmptyString,
} from './ValidationTools';

import {
    ARC,
    BACKGROUNDCOLOR,
    BAR,
    BARTHICKNESS,
    BORDERCOLOR,
    BORDERRADIUS,
    BORDERSKIPPED,
    BORDERWIDTH,
    CHART_CONFIG_VALIDATE,
    CONTEXT,
    DATA,
    DATASETS,
    DISPLAY,
    DRAWACTIVEELEMENTSONTOP,
    ELEMENTS,
    FILL,
    HEIGHT,
    INDEXAXIS,
    LABEL,
    LABELS,
    LEGEND,
    LINE,
    ORI_HORIZONTLE,
    ORI_VERTICLE,
    OPTIONS,
    PLUGINS,
    POINTBACKGROUNDCOLOR,
    POINTBORDERWIDTH,
    POINTHITRADIUS,
    POINTHOVERRADIUS,
    POINTRADIUS,
    POSITION,
    POSITION_TOP,
    POSITION_BOTTOM,
    POSITION_LEFT,
    POSITION_RIGHT,
    RESPONSIVE,
    SCALES,
    STACKED,
    STYLE,
    TEXT,
    TITLE,
    TYPE,
    WIDTH,
    X,
    Y,
} from '../constants/ChartConstants';

const Validator = {
    // Descend down the configuration and assert each node and leave follows some rule.
    validate(ChartConfig) {
        try{
            const { data, options, style, type } = ChartConfig;
            Validator.validateData(data);
            Validator.validateOptions(options);
            Validator.validateStyle(style);
            assertNonEmptyString(type, TYPE, CHART_CONFIG_VALIDATE);    
        } catch(error) {
            assertIsValid(false, 'ChartConfig', error.message);
        }
    },

    validateData(data) {
        try {
            assertIsDefinedNotNull(data[LABELS], LABELS, CONTEXT);
            assertIsDefinedNotNull(data[DATASETS], DATASETS, CONTEXT);
        
            Object.keys(data).forEach(configKey => {
                const configValue = data[configKey];
                switch(configKey) {
                    case LABELS:
                        assertIsStringArray(configValue, configKey, CONTEXT);
                        break;
                    case DATASETS:
                        Validator.validateDatasets(configValue);
                        break;
                    default:
                        assertIsSupported(false, configKey, CONTEXT);
                }
            });
        }
        catch(error) {
            assertIsValid(false, 'Data', error.message);
        }
    },

    validateDatasets(datasets) {
        datasets.forEach(set => {
            assertIsDefinedNotNull(set[DATA], DATA, DATASETS);
            assertIsDefinedNotNull(set[LABEL], LABEL, DATASETS);
            Object.keys(set).forEach(optionKey => {
                const optionValue = set[optionKey];
                switch(optionKey) {
                    case DATA:
                        assertIsNumericArray(optionValue, optionKey, DATASETS);
                        break;
                    case DRAWACTIVEELEMENTSONTOP:
                    case FILL:
                        assertIsBoolean(optionValue, optionKey, DATASETS);
                        break;
                    case BACKGROUNDCOLOR:
                    case BORDERCOLOR:
                    case POINTBACKGROUNDCOLOR:
                        if(Array.isArray(optionValue)) {
                            assertIsStringArray(optionValue, optionKey, DATASETS);
                            break;
                        }
                        // else it could be a string so let fall through.
                    case LABEL:
                    case TYPE:
                        assertIsString(optionValue, optionKey, DATASETS);
                        break;
                    case BARTHICKNESS:
                    case BORDERRADIUS:
                    case BORDERWIDTH:
                    case POINTBORDERWIDTH:
                    case POINTHITRADIUS:
                    case POINTHOVERRADIUS:
                    case POINTRADIUS:
                        assertIsNumber(optionValue, optionKey, DATASETS);
                        break;
                    default:
                        assertIsSupported(false, optionKey, DATASETS);
                }
            });
        });
    },

    validateOptions(options) {
        try{
            assertIsDefinedNotNull(options, OPTIONS, 'validateOptions()');
            const { indexAxis, elements, plugins, responsive, scales } = options;

            if(indexAxis) {
                switch(indexAxis) {
                    case ORI_HORIZONTLE:
                    case ORI_VERTICLE:
                        break;
                    default:
                        assertIsSupported(false, indexAxis, INDEXAXIS);            
                }
            }
        
            if(elements) {
                Validator.validateElements(elements);
            }
        
            if(plugins) {
                Validator.validatePlugins(plugins);
            }
        
            if(responsive) {
                assertIsBoolean(responsive, RESPONSIVE, OPTIONS);
            }
        
            if(scales) {
                Validator.validateScales(scales);
            }    
        } catch(error) {
            assertIsValid(false, OPTIONS, error.message);
        }
    },

    validateScales(scales) {
        Object.keys(scales).forEach(scaleKey => {
            const scale = scales[scaleKey];
            switch(scaleKey) {
                case X:
                case Y:
                    Validator.validateScale(scale);
                    break;
                default:
                    assertIsSupported(false, scaleKey, SCALES);
            }
        });
    },

    validateScale(scale) {
        Object.keys(scale).forEach(config => {
            const configVal = scale[config];
            switch(config) {
                case STACKED:
                    assertIsBoolean(configVal, config, SCALES);
                    break;
                case BARTHICKNESS:
                    assertIsNumber(configVal, config, SCALES);
                    break;
                default:
                    assertIsSupported(false, config, SCALES);
            }
        });
    },

    validateElements(elements) {
        Object.keys(elements).forEach( elementKey => {
            let element = elements[elementKey];
            switch(elementKey) {
                case LINE:
                case BAR:
                case ARC:
                    Validator.validateElement(element, elementKey);
                    break;
                default:
                    assertIsSupported(false, elementKey, ELEMENTS);
            }
        });
    },

    validateElement(element, elementKey) {
        Object.keys(element).forEach(option => {
            let optionVal = element[option];
            switch(option) {
                case BORDERWIDTH:
                case BORDERRADIUS:
                    assertIsNumber(optionVal, option, elementKey);
                    break;
                case BORDERSKIPPED:
                    assertIsBoolean(optionVal, option, elementKey);
                    break;
                default:
                    assertIsSupported(false, option, elementKey);
            }
        });
    },

    validatePlugins(plugins) {
        Object.keys(plugins).forEach(pluginKey => {
            const plugin = plugins[pluginKey];
            switch(pluginKey) {
                case LEGEND:
                case TITLE:
                    Validator.validatePlugin(plugin, pluginKey)
                    break;
                default:
                    assertIsSupported(false, pluginKey, PLUGINS);
            }
        });
    },

    validatePlugin(plugin, pluginKey) {
        Object.keys(plugin).forEach(option => {
            const optionValue = plugin[option];
            switch(option) {
                case DISPLAY:
                    assertIsBoolean(optionValue, option, pluginKey);
                    break;
                case POSITION:
                    switch(optionValue) {
                        case POSITION_BOTTOM:
                        case POSITION_LEFT:
                        case POSITION_RIGHT:
                        case POSITION_TOP:
                                break;
                        default:
                            assertIsSupported(false, optionValue, option);
                    }
                    break;
                case TEXT:
                    assertIsString(optionValue, option, pluginKey);
                    break;
                default:
                    assertIsSupported(false, option, pluginKey);
            }
        });
    },

    validateStyle(style) {
        try{
            Object.keys(style).forEach(config => {
                const configValue = style[config];
                switch(config) {
                    case HEIGHT:
                    case WIDTH:
                        assertIsNumber(configValue, config, STYLE);
                        break;
                    default:
                        assertIsSupported(false, config, STYLE);
                }
            });    
        } catch(error) {
            assertIsValid(false, STYLE, error.message);
        }
    }
}

export default Validator;