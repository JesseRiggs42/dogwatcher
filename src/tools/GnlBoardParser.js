function parseData(htmlLeaderBoardData){
    if(htmlLeaderBoardData === null
        || typeof(htmlLeaderBoardData) != 'string'
        || htmlLeaderBoardData.length === 0) {
        throw new Error("Error: htmlLeaderBoardData must be non-empty string.")
    }

}