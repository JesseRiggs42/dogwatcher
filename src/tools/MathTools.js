function sumArray(scoresArray) {
    
    let scoresTotal = 0;
    scoresArray.forEach(score => {
        scoresTotal += score;
    });

    return scoresTotal;
}

function calculateAverage(scoresArray) {

    let gamesAverage;
    scoresArray.length == 0
        ? gamesAverage = 0
        : gamesAverage = sumArray(scoresArray) / scoresArray.length;
    
    return gamesAverage;
}

function calculateStandardDeviation(scoresArray){

    let average = calculateAverage(scoresArray);
    let squareDiff = 0;
    scoresArray.forEach(score => {
            squareDiff += (score - average) ** 2;
    });
    
    let gamesStandardDeviation;
    scoresArray.length === 0
        ? gamesStandardDeviation = 0
        : gamesStandardDeviation = Math.sqrt(squareDiff / scoresArray.length);

    return gamesStandardDeviation;
}

export { calculateAverage, calculateStandardDeviation, sumArray }