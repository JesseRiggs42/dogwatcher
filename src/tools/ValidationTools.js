function assert(valid, message) {
    if(!valid) {
        throw new Error(message);
    }
}

function assertIsBoolean(boolean, elementName, context) {
    assert(boolean !== null, `Element "${elementName}" in "${context}" is expected to be non-null`);
    assert(typeof(boolean) === 'boolean', `Element "${elementName}" in "${context}" is expected to be of type "boolean" but is "${typeof(boolean)}".`);
}

function assertNonEmptyString(string, elementName, context) {
    assertIsString(string, elementName, context);
    assert(string !== '', `Element "${elementName}" cannot be empty.`);
}

function assertIsDefinedNotNull(element, elementName, context) {
    assert(typeof(element) !== 'undefined', `Element "${elementName}" in "${context}" is expected to be defined.`);
    assert(element !== null, `Element "${elementName}" in "${context}" is expected to be non-null`);
}

function assertIsFunction(func, functionName, context) {
    assertIsDefinedNotNull(func, functionName, context);
    assert(typeof(func) === 'function', `"${functionName}" is expected to be of type "function" but is "${typeof(func)}".`);
}

function assertIsNumber(number, numberName, context) {
    assertIsDefinedNotNull(number, numberName, context);
    assert(!isNaN(Number.parseInt(number)), `"${numberName}" is expected to be a number in "${context}".`);
}

function assertIsNumericArray(numbers, arrayName, context) {
    assertIsDefinedNotNull(numbers, arrayName, context);
    assert(Array.isArray(numbers), `Numeric array "${arrayName}" in "${context}" must be array.`);
    numbers.forEach(number => {
        assert(!isNaN(Number.parseFloat(number)), `Element of numeric array "${arrayName}" in "${context}" must be a number.`);
    });
}

function assertIsStringArray(stringArray, arrayName, context) {
    assertIsDefinedNotNull(stringArray, arrayName, context);
    assert(Array.isArray(stringArray), `Element "${arrayName}" in "${context}" is expected to be of type "array" but is "${typeof(stringArray)}".`);
    stringArray.forEach(string => {
        assert(typeof(string) === 'string', `Element "${string}" of String Array "${arrayName}" in "${context}" is expected to be of type "string" but is "${typeof(string)}".`);
    });
}

function assertIsString(string, stringName, context) {
    assertIsDefinedNotNull(string, stringName, context);
    assert(typeof(string) === 'string', `"${stringName}" in "${context}" is expected to be of type "string" but is "${typeof(string)}".`);
}

function assertIsSupported(supported, optionName, context) {
    if(!supported) {
        throw new Error(`Error: "${optionName}" is not supported by "${context}".`);
    }
}

function assertIsValid(valid, subject, description) {
    if(!valid) {
        throw new Error(`Could not validate "${subject}": ${description}`);
    }
}

function assertNotInstantiated(object, objName) {
    assert(object === undefined || object === null, `Error: "${objName}" cannot be instantiated twice.`);
}

export {
    assert,
    assertIsBoolean,
    assertIsDefinedNotNull,
    assertIsFunction,
    assertIsNumber,
    assertIsNumericArray,
    assertIsString,
    assertIsStringArray,
    assertIsSupported,
    assertIsValid,
    assertNonEmptyString,
    assertNotInstantiated
}
