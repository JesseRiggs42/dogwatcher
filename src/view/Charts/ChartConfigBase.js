import { plugins } from 'chart.js';
import {
    assertIsString,
    assertIsStringArray,
} from '../../tools/ValidationTools';
import {
    BARTHICKNESS,
    BORDERRADIUS,
    BORDERSKIPPED,
    BORDERWIDTH,
    DATA,
    DATASETS,
    DISPLAY,
    ELEMENTS,
    HEIGHT,
    INDEXAXIS,
    LABELS,
    LEGEND,
    ORI_HORIZONTLE,
    ORI_VERTICLE,
    OPTIONS,
    PLUGINS,
    POSITION,
    POSITION_TOP,
    RESPONSIVE,
    SCALES,
    STACKED,
    STYLE,
    TEXT,
    TITLE,
    TYPE,
    WIDTH,
    X,
    Y,
} from '../../constants/ChartConstants';

export default class ChartConfigBase {
    // colors as array https://stackoverflow.com/questions/25594478/different-color-for-each-bar-in-a-bar-chart-chartjs
    constructor(title, type, labels, width, height, isHorizontle) {
        this.datasets = [];
        this.height = height;
        this.horizontle = isHorizontle;
        this.labels = labels;
        this.responsive = true;
        this.title = title;
        this.type = type;
        this.width = width;

        assertIsStringArray(labels, LABELS, this.constructor.name);
        assertIsString(title, TITLE, this.constructor.name);
    }

    // This is the final step. Unless overridden, this will return the same structure of agregated data.
    // The idea being that all decorations will return their own parts concatenated with the parts of their decorated.
    // This should only be called from the highest level.
    getConfig() {
        return {
            [TYPE]    : this.getType(),
            [OPTIONS] : this.getOptions(),
            [DATA]    : this.getData(),
            [STYLE]   : this.getStyle(),
        }
    }

    getData() {
        return {
            [LABELS]  : this.getLabels(),
            [DATASETS]: this.getDatasets(),
        };
    }

    getDatasets() {
        return this.datasets;
    }

    getElements() {
        return {
            bar: {
                [BORDERWIDTH]  : 2,
                [BORDERSKIPPED]: false,
                [BORDERRADIUS] : Number.MAX_VALUE,
            }
        };
    }

    getHeight() {
        return this.height;
    }

    getOptions() {
        return {
            [INDEXAXIS]   : this.isHorizontle() ? ORI_HORIZONTLE : ORI_VERTICLE,
            [ELEMENTS]    : this.getElements(),
            [PLUGINS]     : this.getPlugins(),
            [RESPONSIVE]  : this.isResponsive(),
            [SCALES]      : this.getScales()
        };
    }

    getLabels() {
        return this.labels;
    }

    getPlugins() {
        return {
            [LEGEND]: {
                [POSITION]: POSITION_TOP,
            },
            [TITLE]: {
                [DISPLAY] : true,
                [TEXT]    : this.title
            }
        };
    }

    getScales() {
        return {
            [X]: {
                [STACKED]: true,
            },
            [Y]: {
                [BARTHICKNESS]: 0.1,
                [STACKED]     : true,
            }
        }
    }

    getStyle() {
        return {
            [WIDTH]   : this.getWidth() ?? 750,
            [HEIGHT]  : this.getHeight() ?? 350,
        };
    }
    
    getTitle() {
        return this.title;
    }

    getType() {
        return this.type;
    }

    getWidth() {
        return this.width;
    }

    isHorizontle() {
        return this.horizontle;
    }

    isResponsive() {
        return this.responsive;
    }

}
