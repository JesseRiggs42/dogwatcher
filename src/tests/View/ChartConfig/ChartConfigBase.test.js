import ChartConfigBase from "../../../view/Charts/ChartConfigBase";
import { validate } from '../../../tools/ChartConfigValidator';

describe('ChartConfigBase behaves as expected', () => {
    describe('with valid parameters', () => {
        let title = 'Config Title';
        let type = 'Config Type';
        let labels = ['label1','label2'];
        let width = 10;
        let height = 20;
        let isHorizontle = 'false';

        let chartConfigBase;
        test('constructor() returns new ChartConfigBase.', () => {
            chartConfigBase = new ChartConfigBase(title, type, labels, width, height, isHorizontle);
        });
        test('getDatasets() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getDatasets())).toMatchSnapshot();
        });
        test('getElements() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getElements())).toMatchSnapshot();
        });
        test('getHeight() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getHeight())).toMatchSnapshot();
        });
        test('getOptions() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getOptions())).toMatchSnapshot();
        });
        test('getLabels() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getLabels())).toMatchSnapshot();
        });
        test('getPlugins() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getPlugins())).toMatchSnapshot();
        });
        test('getScales() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getScales())).toMatchSnapshot();
        });
        test('getStyle() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getStyle())).toMatchSnapshot();
        });
        test('getTitle() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getTitle())).toMatchSnapshot();
        });
        test('getType() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getType())).toMatchSnapshot();
        });
        test('getWidth() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getWidth())).toMatchSnapshot();
        });
        test('isHorizontle() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.isHorizontle())).toMatchSnapshot();
        });
        test('isResponsive() matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.isResponsive())).toMatchSnapshot();
        });
        test('validate() returns normally.', () => {
            validate(chartConfigBase.getConfig());
        });
        test('getConfig()  matches snapshot.', () => {
            expect(JSON.stringify(chartConfigBase.getConfig())).toMatchSnapshot();
        });
    });
    describe('with invalid parameters', () => {
        const title = 'Title';
        const type = 'Type';
        const labels = ['lable'];
        const width = 10;
        const height = 20;
        const isHorizontle = false;
        test('throws on undefined title.', () => {
            try{
                new ChartConfigBase(undefined, type, labels, width, height, isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on non-string title.', () => {
            try{
                new ChartConfigBase(1234, type, labels, width, height, isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on undefined type.', () => {
            try{
                new ChartConfigBase(title, undefined, labels, width, height, isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on non-string type.', () => {
            try{
                new ChartConfigBase(title, 1234, labels, width, height, isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on non-boolean horizontle.', () => {
            try{
                new ChartConfigBase(title, type, labels, width, height, 'true');
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on invalid labels.', () => {
            try{
                new ChartConfigBase(title, type, [1,2], width, height, isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on non-numeric width.', () => {
            try{
                new ChartConfigBase(title, type, labels, 'width', height, isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
        test('throws on non-numeric height.', () => {
            try{
                new ChartConfigBase(title, type, labels, width, 'height', isHorizontle);
                expect(false).toEqual('Should have thrown.');
            } catch(error) {
                expect(error.message).toMatchSnapshot();
            }
        });
    });
});