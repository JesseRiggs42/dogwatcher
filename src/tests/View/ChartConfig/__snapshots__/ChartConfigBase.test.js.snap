// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartConfigBase behaves as expected with invalid parameters throws on invalid labels. 1`] = `"Element \\"1\\" of String Array \\"labels\\" in \\"ChartConfigBase\\" is expected to be of type \\"string\\" but is \\"number\\"."`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on non-boolean horizontle. 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

Expected: [32m\\"Should have thrown.\\"[39m
Received: [31mfalse[39m"
`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on non-numeric height. 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

Expected: [32m\\"Should have thrown.\\"[39m
Received: [31mfalse[39m"
`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on non-numeric width. 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

Expected: [32m\\"Should have thrown.\\"[39m
Received: [31mfalse[39m"
`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on non-string title. 1`] = `"\\"title\\" in \\"ChartConfigBase\\" is expected to be of type \\"string\\" but is \\"number\\"."`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on non-string type. 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

Expected: [32m\\"Should have thrown.\\"[39m
Received: [31mfalse[39m"
`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on undefined title. 1`] = `"Element \\"title\\" in \\"ChartConfigBase\\" is expected to be defined."`;

exports[`ChartConfigBase behaves as expected with invalid parameters throws on undefined type. 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

Expected: [32m\\"Should have thrown.\\"[39m
Received: [31mfalse[39m"
`;

exports[`ChartConfigBase behaves as expected with valid parameters getConfig()  matches snapshot. 1`] = `"{\\"type\\":\\"Config Type\\",\\"options\\":{\\"indexAxis\\":\\"y\\",\\"elements\\":{\\"bar\\":{\\"borderWidth\\":2,\\"borderSkipped\\":false,\\"borderRadius\\":1.7976931348623157e+308}},\\"plugins\\":{\\"legend\\":{\\"position\\":\\"top\\"},\\"title\\":{\\"display\\":true,\\"text\\":\\"Config Title\\"}},\\"responsive\\":true,\\"scales\\":{\\"x\\":{\\"stacked\\":true},\\"y\\":{\\"barThickness\\":0.1,\\"stacked\\":true}}},\\"data\\":{\\"labels\\":[\\"label1\\",\\"label2\\"],\\"datasets\\":[]},\\"style\\":{\\"width\\":10,\\"height\\":20}}"`;

exports[`ChartConfigBase behaves as expected with valid parameters getDatasets() matches snapshot. 1`] = `"[]"`;

exports[`ChartConfigBase behaves as expected with valid parameters getElements() matches snapshot. 1`] = `"{\\"bar\\":{\\"borderWidth\\":2,\\"borderSkipped\\":false,\\"borderRadius\\":1.7976931348623157e+308}}"`;

exports[`ChartConfigBase behaves as expected with valid parameters getHeight() matches snapshot. 1`] = `"20"`;

exports[`ChartConfigBase behaves as expected with valid parameters getLabels() matches snapshot. 1`] = `"[\\"label1\\",\\"label2\\"]"`;

exports[`ChartConfigBase behaves as expected with valid parameters getOptions() matches snapshot. 1`] = `"{\\"indexAxis\\":\\"y\\",\\"elements\\":{\\"bar\\":{\\"borderWidth\\":2,\\"borderSkipped\\":false,\\"borderRadius\\":1.7976931348623157e+308}},\\"plugins\\":{\\"legend\\":{\\"position\\":\\"top\\"},\\"title\\":{\\"display\\":true,\\"text\\":\\"Config Title\\"}},\\"responsive\\":true,\\"scales\\":{\\"x\\":{\\"stacked\\":true},\\"y\\":{\\"barThickness\\":0.1,\\"stacked\\":true}}}"`;

exports[`ChartConfigBase behaves as expected with valid parameters getPlugins() matches snapshot. 1`] = `"{\\"legend\\":{\\"position\\":\\"top\\"},\\"title\\":{\\"display\\":true,\\"text\\":\\"Config Title\\"}}"`;

exports[`ChartConfigBase behaves as expected with valid parameters getScales() matches snapshot. 1`] = `"{\\"x\\":{\\"stacked\\":true},\\"y\\":{\\"barThickness\\":0.1,\\"stacked\\":true}}"`;

exports[`ChartConfigBase behaves as expected with valid parameters getStyle() matches snapshot. 1`] = `"{\\"width\\":10,\\"height\\":20}"`;

exports[`ChartConfigBase behaves as expected with valid parameters getTitle() matches snapshot. 1`] = `"\\"Config Title\\""`;

exports[`ChartConfigBase behaves as expected with valid parameters getType() matches snapshot. 1`] = `"\\"Config Type\\""`;

exports[`ChartConfigBase behaves as expected with valid parameters getWidth() matches snapshot. 1`] = `"10"`;

exports[`ChartConfigBase behaves as expected with valid parameters isHorizontle() matches snapshot. 1`] = `"\\"false\\""`;

exports[`ChartConfigBase behaves as expected with valid parameters isResponsive() matches snapshot. 1`] = `"true"`;
